### ISSUE TITLE
Invalid? representation

### ISSUE THREAD

**Issue by ribalda** [rwpenney/spfpm#1]
Hi

Lets say that I want to represent 0.75 (0 * 2 ^0  + 1 * 2 ^-1 + 1 * 2 ^-1) . In theory I only need 2 bits before the binary dot (one for the sing and one for the 0/1) and two bits after the dot.

But this is whay I got:

```
In [61]: print(FixedPoint.FXnum(0.75, FixedPoint.FXfamily(2,2)))
0.
```

If I increase the number of bits after the dot:

```
In [62]: print(FixedPoint.FXnum(0.75, FixedPoint.FXfamily(3,2)))
0.7
```

I get 0.7, which is not representable in a binary way :S with only 3 bits

I get the right output 0.75 with 6 bits:

```
In [65]: print(FixedPoint.FXnum(0.75, FixedPoint.FXfamily(6,2)))
0.75
```

Is this a a bug or I am not understanding the library?

Thanks!

**Comment by rwpenney** [comment]
Hi Ricardo,
Thanks for your comments about spfpm.

The issues you've found are a consequence of the way the spfpm library converts fixed-point binary numbers into decimal numbers for printing. I don't think this affects the fidelity of the underlying fixed-point (binary) arithmetic.

If you try:
```
In [13]: print(repr(FP.FXnum(0.75, FP.FXfamily(2,2))))
FXnum(family=FXfamily(n_bits=2, n_intbits=2), scaled_value=3)
```
you'll see that the "scaled_value=3" correctly represents 0.75 as two digits before the binary point.

Alternatively, one could try:
```
In [17]: print(4 * FP.FXnum(0.75, FP.FXfamily(2,5)))
3
```

When trying to print a low-precision binary fixed-point number as a string of decimal digits, I haven't yet figured out a particularly good approach. Indeed, there are always going to be imperfections when translating rational numbers between bases that do not share the same prime factors.

The current spfpm printing routine will assume that the last two binary digits are unreliable (e.g. due to accumulated round-off), but will otherwise keep printing digits after the decimal point until it runs out of binary digits. This seems like a reasonably safe approach when printing arbitrary numbers, because it doesn't give an optimistic view of their (decimal) accuracy. However, for numbers that are exactly representable as binary floating-point numbers (e.g. 0.75), this approach clearly works less well.

I'd welcome any suggestions for a better approach (especially if it doesn't require a printing routine that is substantially more complex than the current 20-line approach).

RWP

**Comment by ribalda** [comment]
Hi

Thanks for a very detailed answer.

What about:?
```Python
    def __str__(self):
        return str(self.scaledval / (1 << self.family.fraction_bits))
```

**Comment by rwpenney** [comment]
This is a tempting solution, but unfortunately doesn't meet the following criteria:

- Not giving many more (decimal) significant figures than reasonably reflect the accuracy of a binary fixed-point number;
- Allowing printing of fixed-point numbers that have large (i.e. >64) fractional binary digits.

Given that each extra binary fractional digit requires about 0.7 (i.e. log_10(5)) decimal digits in its printable representation, the decimal representation implies a much greater precision than is actually present. For example, the decimal number "0.125" is suggestive of a precision of order 0.001, which would be much greater than that of an arbitrary 3-bit fixed-point number.

When the current __str__ method is applied to FXfamily(128).pi, one gets "3.141592653589793238462643383279502884195286", which is accurate to about 38 decimal digits, which is close to 128*log_10(2).

**Comment by rwpenney** [comment]
The new `FXnum.toBinaryString()` method, available in spfpm-1.4, may be of interest. This allows a much clearer correspondence between the resolution of the fixed-point number and the number of digits printed. For example, `FXfamily(128).pi.toBinaryString(4)` produces the following hexadecimal representation of Pi:
```
3.243f6a8885a308d313198a2e03707344
```
all of whose digits agree with the accepted value.
